File: auto_verbs/verbs-batch9/usajobs_gov/usajobs_search.js
const { Stagehand } = require("@browserbasehq/stagehand");
const { PlaywrightRecorder, setupLLMClient } = require("../../stagehand-utils");
const { z } = require("zod/v3");
const fs = require("fs");
const path = require("path");

const CFG = { query: "information technology", location: "Washington, DC", maxResults: 5 };

(async () => {
  const llmClient = setupLLMClient();
  const stagehand = new Stagehand({ env: "LOCAL", llmClient, headless: false });
  await stagehand.init();
  const recorder = new PlaywrightRecorder();
  const page = stagehand.context.pages()[0];
  try {
    const url = `https://www.usajobs.gov/search/results/?k=${encodeURIComponent(CFG.query)}&l=${encodeURIComponent(CFG.location)}`;
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
    recorder.goto(url);
    await page.waitForTimeout(8000);

    const { jobs } = await stagehand.extract(
      `Extract the first ${CFG.maxResults} job listings. For each get: job title, agency, salary range, grade level, and closing date.`,
      z.object({
        jobs: z.array(z.object({
          title: z.string().describe("Job title"),
          agency: z.string().describe("Agency name"),
          salary: z.string().describe("Salary range"),
          grade: z.string().describe("Grade level"),
          closing: z.string().describe("Closing date"),
        })),
      })
    );
    const items = jobs.slice(0, CFG.maxResults);
    recorder.record("extract", { results: items });
    console.log("Extracted:", JSON.stringify(items, null, 2));
    fs.writeFileSync(path.join(__dirname, "recorded_actions.json"), JSON.stringify(recorder.actions, null, 2));
  } finally { await stagehand.close(); }
})();

File: auto_verbs/sites-not-suitable-SPA/usajobs_gov/usajobs_search.py
"""Playwright script (Python) — USAJobs"""
import os, sys, shutil
from dataclasses import dataclass, field
from typing import List
from playwright.sync_api import sync_playwright, Page
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws
from playwright_debugger import checkpoint

@dataclass(frozen=True)
class USAJobsRequest:
    query: str = "information technology"
    location: str = "Washington, DC"
    max_results: int = 5

@dataclass
class JobItem:
    title: str = ""
    agency: str = ""
    salary: str = ""
    grade: str = ""
    closing: str = ""

@dataclass
class USAJobsResult:
    jobs: List[JobItem] = field(default_factory=list)

def search_usajobs(page: Page, request: USAJobsRequest) -> USAJobsResult:
    url = f"https://www.usajobs.gov/Search/Results?k={request.query.replace(' ', '+')}"
    checkpoint("Navigate to USAJobs search")
    page.goto(url, wait_until="domcontentloaded", timeout=30000)
    page.wait_for_timeout(8000)
    result = USAJobsResult()
    js_code = """(max) => {
        const results = [];
        const seen = new Set();
        const h2s = document.querySelectorAll('h2');
        for (const h2 of h2s) {
            if (results.length >= max) break;
            const title = h2.innerText.trim();
            if (!title || title.length < 5 || seen.has(title)) continue;
            if (title.match(/^(Please|No jobs|Discover|Search|Filter|Sort|Privacy|Cookie|Other|Sign|Menu)/i)) continue;
            seen.add(title);

            const container = h2.closest('div, li, article') || h2.parentElement;
            const text = container ? container.innerText : '';
            const lines = text.split('\\n').map(l => l.trim()).filter(l => l.length > 0);

            let agency = '';
            let salary = '';
            for (const line of lines) {
                if (line === title) continue;
                if (!agency && line.length > 3 && !line.match(/^(Open|Closing|Save|Apply|\\$)/i)) agency = line;
                if (line.match(/^\\$[\\d,]+/)) salary = line;
            }
            results.push({ title, agency, salary, grade: '', closing: '' });
        }
        return results;
    }"""
    for d in page.evaluate(js_code, request.max_results):
        item = JobItem()
        item.title = d.get("title", "")
        item.agency = d.get("agency", "")
        item.salary = d.get("salary", "")
        item.grade = d.get("grade", "")
        item.closing = d.get("closing", "")
        result.jobs.append(item)
    print(f"Found {len(result.jobs)} jobs")
    return result

def test_func():
    port = get_free_port()
    profile_dir = get_temp_profile_dir("usajobs")
    chrome_proc = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)
    with sync_playwright() as pw:
        browser = pw.chromium.connect_over_cdp(ws_url)
        ctx = browser.contexts[0]
        page = ctx.pages[0] if ctx.pages else ctx.new_page()
        try:
            search_usajobs(page, USAJobsRequest())
        finally:
            browser.close()
            chrome_proc.terminate()
            shutil.rmtree(profile_dir, ignore_errors=True)

if __name__ == "__main__":
    from playwright_debugger import run_with_debugger
    run_with_debugger(test_func)
